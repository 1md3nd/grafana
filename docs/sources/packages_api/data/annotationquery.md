+++
# -----------------------------------------------------------------------
# Do not edit this file. It is automatically generated by API Documenter.
# -----------------------------------------------------------------------
title = "AnnotationQuery"
keywords = ["grafana","documentation","sdk","@grafana/data"]
type = "docs"
disable_edit_link = true
+++

## AnnotationQuery interface

This JSON object is stored in the dashboard json model.

<b>Signature</b>

```typescript
export interface AnnotationQuery<TQuery extends DataQuery = DataQuery> 
```
<b>Import</b>

```typescript
import { AnnotationQuery } from '@grafana/data';
```
<b>Properties</b>

|  Property | Type | Description |
|  --- | --- | --- |
|  [datasource](#datasource-property) | <code>string</code> |  |
|  [enable](#enable-property) | <code>boolean</code> |  |
|  [hide](#hide-property) | <code>boolean</code> |  |
|  [iconColor](#iconcolor-property) | <code>string</code> |  |
|  [mappings](#mappings-property) | <code>AnnotationEventMappings</code> |  |
|  [name](#name-property) | <code>string</code> |  |
|  [target](#target-property) | <code>TQuery</code> |  |

### datasource property

<b>Signature</b>

```typescript
datasource: string;
```

### enable property

<b>Signature</b>

```typescript
enable: boolean;
```

### hide property

<b>Signature</b>

```typescript
hide?: boolean;
```

### iconColor property

<b>Signature</b>

```typescript
iconColor: string;
```

### mappings property

<b>Signature</b>

```typescript
mappings?: AnnotationEventMappings;
```

### name property

<b>Signature</b>

```typescript
name: string;
```

### target property

<b>Signature</b>

```typescript
target?: TQuery;
```
