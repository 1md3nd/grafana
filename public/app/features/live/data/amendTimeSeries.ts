import { closestIdx } from "./StreamingDataFrame";

// prevData and nextData are assumed sorted ASC on reference [0] arrays
// nextData is assumed to be contiguous, only edges are checked for overlap
// ...so prev: [1,2,5] + next: [3,4,6] -> [1,2,3,4,6]
export function amendTimeSeries(prevData: [number[], any[]], nextData: [number[], any[]]): [number[], any[]] {
  let [pr, pv] = prevData ?? [[], []];
  let [nr, nv] = nextData ?? [[], []];

  let pLen = pr.length;
  let pStart = pr[0];
  let pEnd = pr[pLen - 1];

  let nLen = nr.length;
  let nStart = nr[0];
  let nEnd = nr[nLen - 1];

  let r: number[];
  let v: any[];

  if (pLen) {
    if (nLen) {
      // append, no overlap
      if (nStart > pEnd) {
        //	console.log('append, no overlap');
        r = pr.concat(nr);
        v = pv.concat(nv);
      }
      // prepend, no overlap
      else if (nEnd < pStart) {
        //	console.log('prepend, no overlap');
        r = nr.concat(pr);
        v = nv.concat(pv);
      }
      // full replace
      else if (nStart <= pStart && nEnd >= pEnd) {
        r = nr;
        v = nv;
      }
      // partial replace
      else if (nStart > pStart && nEnd < pEnd) {
      }
      // append, with overlap
      else if (nStart >= pStart) {
        //	console.log('append, with overlap');
        let idx = closestIdx(nStart, pr);
        idx = pr[idx] < nStart ? idx - 1 : idx;
        r = pr.slice(0, idx).concat(nr);
        v = pv.slice(0, idx).concat(nv);
      }
      // prepend, with overlap
      else if (nEnd >= pStart) {
        //	console.log('prepend, with overlap');
        let idx = closestIdx(nEnd, pr);
        idx = pr[idx] < nEnd ? idx : idx + 1;
        r = nr.concat(pr.slice(idx));
        v = nv.concat(pv.slice(idx));
      }
    } else {
      r = pr;
      v = pv;
    }
  } else {
    if (nLen) {
      r = nr;
      v = nv;
    } else {
      // defaulted to empty arrays, just reuse
      r = pr;
      v = pv;
    }
  }

  return [r!, v!];
}

type TimeSeries = [times: number[], vals: any[]];

export function trimTimeSeries(data: TimeSeries, fromTime: number, toTime: number): TimeSeries {
  let [times, vals] = data;
  let fromIdx;
  let toIdx;

  // trim to bounds
  if (times[0] < fromTime) {
    fromIdx = closestIdx(fromTime, times);

    if (times[fromIdx] < fromTime) {
      fromIdx++;
    }
  }

  if (times[times.length - 1] > toTime) {
    toIdx = closestIdx(toTime, times);

    if (times[toIdx] > toTime) {
      toIdx--;
    }
  }

  if (fromIdx != null || toIdx != null) {
    //	console.log('trim!', fromIdx ?? 0, toIdx);
    times = times.slice(fromIdx ?? 0, toIdx);
    vals = vals.slice(fromIdx ?? 0, toIdx);
  }

  return [times, vals];
}
