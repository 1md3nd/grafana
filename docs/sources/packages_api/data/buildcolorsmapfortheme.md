+++
# -----------------------------------------------------------------------
# Do not edit this file. It is automatically generated by API Documenter.
# -----------------------------------------------------------------------
title = "buildColorsMapForTheme"
keywords = ["grafana","documentation","sdk","@grafana/data"]
type = "docs"
+++

## buildColorsMapForTheme() function

### buildColorsMapForTheme() function

<b>Signature</b>

```typescript
export declare function buildColorsMapForTheme(theme: GrafanaTheme): Record<Color, string>;
```
<b>Import</b>

```typescript
import { buildColorsMapForTheme } from '@grafana/data';
```
<b>Parameters</b>

|  Parameter | Type | Description |
|  --- | --- | --- |
|  theme | <code>GrafanaTheme</code> |  |

<b>Returns:</b>

`Record<Color, string>`

