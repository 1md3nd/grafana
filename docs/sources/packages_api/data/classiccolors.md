+++
# -----------------------------------------------------------------------
# Do not edit this file. It is automatically generated by API Documenter.
# -----------------------------------------------------------------------
title = "classicColors"
keywords = ["grafana","documentation","sdk","@grafana/data"]
type = "docs"
+++

## classicColors variable

### classicColors variable

<b>Signature</b>

```typescript
classicColors: string[]
```
<b>Import</b>

```typescript
import { classicColors } from '@grafana/data';
```
