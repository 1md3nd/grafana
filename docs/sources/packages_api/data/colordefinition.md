+++
# -----------------------------------------------------------------------
# Do not edit this file. It is automatically generated by API Documenter.
# -----------------------------------------------------------------------
title = "ColorDefinition"
keywords = ["grafana","documentation","sdk","@grafana/data"]
type = "docs"
disable_edit_link = true
+++

## ColorDefinition type

### ColorDefinition type

<b>Signature</b>

```typescript
export declare type ColorDefinition = {
    hue: Hue;
    isPrimary?: boolean;
    name: Color;
    variants: ThemeVariants;
};
```
<b>Import</b>

```typescript
import { ColorDefinition } from '@grafana/data';
```
