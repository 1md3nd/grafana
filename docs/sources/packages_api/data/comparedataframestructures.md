+++
# -----------------------------------------------------------------------
# Do not edit this file. It is automatically generated by API Documenter.
# -----------------------------------------------------------------------
title = "compareDataFrameStructures"
keywords = ["grafana","documentation","sdk","@grafana/data"]
type = "docs"
disable_edit_link = true
+++

## compareDataFrameStructures() function

> This API is provided as a preview for developers and may change based on feedback that we receive. Do not use this API in a production environment.
> 

### compareDataFrameStructures() function

Returns true if both frames have the same list of fields and configs. Field may have diferent names, labels and values but share the same structure

To compare multiple frames use:

```
compareArrayValues(a, b, framesHaveSameStructure);

```
NOTE: this does a shallow check on the FieldConfig properties, when using the query editor, this should be sufficient, however if applicaitons are mutating properties deep in the FieldConfig this will not recognize a change

<b>Signature</b>

```typescript
export declare function compareDataFrameStructures(a: DataFrame, b: DataFrame, skipProperties?: string[]): boolean;
```
<b>Import</b>

```typescript
import { compareDataFrameStructures } from '@grafana/data';
```
<b>Parameters</b>

|  Parameter | Type | Description |
|  --- | --- | --- |
|  a | <code>DataFrame</code> |  |
|  b | <code>DataFrame</code> |  |
|  skipProperties | <code>string[]</code> |  |

<b>Returns:</b>

`boolean`

