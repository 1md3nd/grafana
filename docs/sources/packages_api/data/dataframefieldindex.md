+++
# -----------------------------------------------------------------------
# Do not edit this file. It is automatically generated by API Documenter.
# -----------------------------------------------------------------------
title = "DataFrameFieldIndex"
keywords = ["grafana","documentation","sdk","@grafana/data"]
type = "docs"
disable_edit_link = true
+++

## DataFrameFieldIndex interface

Describes where a specific data frame field is located within a dataset of type DataFrame\[\]

<b>Signature</b>

```typescript
export interface DataFrameFieldIndex 
```
<b>Import</b>

```typescript
import { DataFrameFieldIndex } from '@grafana/data';
```
<b>Properties</b>

|  Property | Type | Description |
|  --- | --- | --- |
|  [fieldIndex](#fieldindex-property) | <code>number</code> |  |
|  [frameIndex](#frameindex-property) | <code>number</code> |  |

### fieldIndex property

<b>Signature</b>

```typescript
fieldIndex: number;
```

### frameIndex property

<b>Signature</b>

```typescript
frameIndex: number;
```
