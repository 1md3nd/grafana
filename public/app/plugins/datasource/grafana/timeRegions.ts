import { TimeRange, DataFrame, FieldType, ArrayVector, getTimeZoneInfo, DataFrameType, DataTopic } from '@grafana/data';
import { calculateTimesWithin } from 'app/core/utils/timeRegions';

import { TimeRegionConfig } from './types';

export function doTimeRegionQuery(
  name: string,
  config: TimeRegionConfig,
  range: TimeRange,
  tz: string
): DataFrame | undefined {
  if (!config) {
    return undefined;
  }
  const regions = calculateTimesWithin(config, range); // UTC
  if (!regions.length) {
    return undefined;
  }

  const times: number[] = [];
  const timesEnd: number[] = [];
  const texts: string[] = [];

  const regionTimezone = config.timezone ?? tz;

  for (const region of regions) {
    let from = region.from;
    let to = region.to;

    const info = getTimeZoneInfo(regionTimezone, from);
    if (info) {
      const offset = info.offsetInMins * 60 * 1000;
      from += offset;
      to += offset;
    }

    times.push(from);
    timesEnd.push(to);
    texts.push(name);
  }

  return {
    meta: {
      type: DataFrameType.TimeRanges,
      //  dataTopic: DataTopic.Annotations,
    },
    fields: [
      { name: 'time', type: FieldType.time, values: new ArrayVector(times), config: {} },
      { name: 'timeEnd', type: FieldType.time, values: new ArrayVector(timesEnd), config: {} },
      { name: 'text', type: FieldType.string, values: new ArrayVector(texts), config: {} },
    ],
    length: times.length,
  };
}
