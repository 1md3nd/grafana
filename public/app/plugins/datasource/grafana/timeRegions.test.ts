import { dateTime, toDataFrameDTO } from '@grafana/data';

import { doTimeRegionQuery } from './timeRegions';

describe('grafana data source', () => {
  it('supports time region query', () => {
    const out = doTimeRegionQuery(
      { name: 'T1', color: 'green', fromDayOfWeek: 1, toDayOfWeek: 2, line: true },
      {
        from: dateTime('2021-01-00', 'YYYY-MM-dd'),
        to: dateTime('2021-02-00', 'YYYY-MM-dd'),
        raw: {
          to: '',
          from: '',
        },
      },
      'utc'
    );

    expect(toDataFrameDTO(out!)).toMatchSnapshot(); // inline??
  });
});
