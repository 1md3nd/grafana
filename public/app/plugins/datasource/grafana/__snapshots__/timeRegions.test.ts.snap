// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grafana data source supports time region query 1`] = `
{
  "fields": [
    {
      "config": {},
      "labels": undefined,
      "name": "time",
      "type": "time",
      "values": [
        1678060800000,
        1678665600000,
        1679270400000,
        1679875200000,
      ],
    },
    {
      "config": {},
      "labels": undefined,
      "name": "timeEnd",
      "type": "time",
      "values": [
        1678233599000,
        1678838399000,
        1679443199000,
        1680047999000,
      ],
    },
    {
      "config": {},
      "labels": undefined,
      "name": "text",
      "type": "string",
      "values": [
        "test",
        "test",
        "test",
        "test",
      ],
    },
  ],
  "meta": {
    "type": "timeranges",
  },
  "name": undefined,
  "refId": undefined,
}
`;
